### PR TITLE
Fix rotation calculations. Add tests.

### PR DESCRIPTION
**Pull request by TimAbraldes**
The j2me API has far fewer footguns than the HTML Canvas API.
Implementing it using the HTML Canvas API involves a lot of severed
feet.

WIP because tests are failing

### CODE REVIEW

**Comment by TimAbraldes** [comment]
Most tests are now fixed. The remaining issue is that, for some reason, strings are drawn with a bigger font

**Comment by TimAbraldes** [comment]
Turns out the font thing was only happening on my machine. Travis was failing due to pixel differences but manual inspection showed that the images were identical-enough-that-I-couldn't-tell-a-difference. Added thresholds based on the pixel differences that Travis reported

**Comment by TimAbraldes** [comment]
No longer WIP

**Comment by marco-c** [comment]
Could you use the red rectangle image instead of the Firefox logo? This should make the number of different pixels much smaller (I think 3000 different pixels are so many that we won't be able to catch regressions).

**Comment by TimAbraldes** [comment]
> Could you use the red rectangle image instead of the Firefox logo? This should make the number of different pixels much smaller (I think 3000 different pixels are so many that we won't be able to catch regressions).

A red rectangle by itself isn't super useful for testing (e.g.) mirrored rotation. I can use a test image that I create on the fly; it will be a rectangle that is larger in one dimension than the other (probably taller than it is wide) with distinctly-colored rectangles in each of its 4 corners

**Comment by TimAbraldes** [comment]
Updated tests to use solid-color images instead of the firefox logo. Let's see how the Travis run turns out!

**Comment by TimAbraldes** [comment]
Hmmm.... TransMirrorRot90AnchorVCenterRight and TransMirrorRot90AnchorBottomRight don't look right

**Comment by marco-c** [comment]
For DrawRegionTransMirrorRot180AnchorVCenterHCenter, this is the diff between the reference image and the image in the Travis build:
diff

Reference image:
image1


Image from the Travis build:
image2

**Comment by TimAbraldes** [comment]
Rebased and fixed some lingering issues. Also added tests for the remaining rotation/anchor combinations. Note that we've passed Travis' log limit of 10,000 lines. Not sure what to do there.

File: midp/gfx.js [hunk 1]
```diff
@@ -1178,38 +1135,114 @@ var currentlyFocusedTextEditor;
...
+            case TRANS_MIRROR_ROT270:
+                w = sh;
+                h = sw;
+                break;
+        }
+
+        // Make `absX` and `absY` the top-left coordinates where we will
+        // place the image in absolute coordinates
+        if (0 != (anchor & HCENTER)) {
+            absX -= (w / 2);
```
Comment (marco-c) [inline midp/gfx.js]: Can you use *(w >>> 1) | 0*? Maybe it's the absence of truncation that is generating the small differences in the rendering?

**Comment by TimAbraldes** [comment]
Added a changeset that switches to ((w >>> 1) | 0) and ((h >>> 1) | 0)

**Comment by marco-c** [comment]
I guess the images need to be re-generated now.

**Comment by TimAbraldes** [comment]
Startup benchmark 100 rounds, 10s round delay:

|      Test     | Baseline Mean |   Mean   |  +/-  |   %   |       P       |    Min   |    Max   | 
|:--------------|--------------:|---------:|------:|------:|--------------:|---------:|---------:|
| startupTime   |       1,641ms |  1,624ms | -17ms | -1.02 | INSIGNIFICANT |  1,537ms |  1,831ms | 
| totalSize     |      28,894kb | 29,022kb | 128kb |  0.44 | INSIGNIFICANT | 27,479kb | 30,968kb | 
| domSize       |         122kb |    122kb |   0kb | -0.00 | INSIGNIFICANT |    122kb |    123kb | 
| styleSize     |         435kb |    435kb |   0kb |  0.08 | INSIGNIFICANT |    435kb |    448kb | 
| jsObjectsSize |      17,371kb | 17,388kb |  17kb |  0.10 |         WORSE | 17,372kb | 17,434kb | 
| jsStringsSize |         666kb |    720kb |  54kb |  8.18 |         WORSE |    719kb |    725kb | 
| jsOtherSize   |       9,936kb |  9,990kb |  54kb |  0.54 | INSIGNIFICANT |  8,460kb | 11,940kb | 
| otherSize     |         363kb |    365kb |   2kb |  0.61 | INSIGNIFICANT |    356kb |    447kb |

**Pull request merged**